Data Governance for the ML Pipeline (#3115)
alan-turing-institute/the-turing-way: book/website/project-design/data-governance/data-gov-ml.md

(pd-dg-ml)=

# Data Governance for the Machine Learning Pipeline

> The Machine Learning (ML) pipeline consists of a series of activities including the collection of data, training of an ML model, and the deployment of the model into use. 

Reviewer: Perhaps A machine learning pipeline?

> Data is integral throughout the ML process and the methods for which data is collected, annotated, processed, and shared will impact individuals who may be the subjects in or creators of the data as well as communities represented in the data.

Reviewer: I'm wondering if perhaps this should be separated out into the subjects (for whom privacy is a concern but not necessarily access) and people who want access to the completed datasets?

This chapter will cover examples of data governance practices for ML for different steps in the pipeline such as: 
- Data Collection
- Data Management
- Data Processing

## Data Collection

> Many ML models are trained using datasets collected by the research team, which may be proprietary, or by using an open dataset that is available for download.

Reviewer: ```suggestion
Many ML models are trained using datasets collected by a research team, which may be proprietary, or by using an open dataset that is available for download (sometimes with restrictions on its usage).
```

Clarified since many models are only open for academic research.

The deep learning (DL) family of models, in particular, rely on massive corpuses of data such as text, code, images, sound, and other media.
The process of data collection depends on the type and volume of data required and sources for acquisition.
Many DL models rely on Internet data due to the sheer volume of digital content that is available on the web. 
For example, [ImageNet datasets](https://www.image-net.org/about.php) are sourced from web images from image hosting websites like Flickr, and [LAION datasets](https://laion.ai/faq/) come from web crawling sources like Common Crawl.

> These methods of data collection through web scraping have raised issues regarding data quality and bias due to the nature of uncurated Internet data, as well as issues of data rights as web scraped data sets often include copyrighted data or data obtained without consent that may be in violation of the data license and terms of use.

Reviewer: ```suggestion
These methods of data collection through web scraping have raised issues regarding data quality and bias due to the nature of using uncurated Internet data, as well as issues of data rights as web scraped data sets often include copyrighted data or data obtained without consent that may be in violation of the data license and terms of use.
```

I'm not sure uncurated is the right word here, since pedantically any information on the internet has been curated (implicitly or explicitly) by someone.

Datasheets are resources accompanying datasets that can provide key information about the motivation to create the dataset, 
the data collection process, the dataset composition and pre-processing, and legal and ethical considerations when using the data.
These resources helps contextualise the resource and is an opportunity for dataset creators to share any concerns, suggestions,
and other advice regarding responsible use of the dataset.
The Datasheets for Datasets paper contains a list of example questions that data creators should ask themselves and document.

## Data Management

Once data collection is complete, challenges emerge around data storage, access, and making changes.
Because state-of-the-art ML datasets are very large, they tend to be stored in a central location, often a cloud provider database, and made available for download under certain circumstances.
In addition to creating a governance structure for the opening or gating access to certain users, there is the challenge of creating a governance structure to allow people to request to opt-out their data from the dataset.

> Because opt-out occurs after a dataset has already been created, it will not remove the data from original versions of the dataset, but can facilitate removal from future versions.

Reviewer: Is this separate from granting rights before the fact? I.e. people who agree to have their data harvested should not be allowed to opt-out in that sense. Additionally, I thought all the data was to be onlywith permission / permissive licensing, so this seems like an oversight / complaint mechanism?

Author: This is separate because after data has been collected, opt-in no longer an option.

Some teams think that data creators should have the option to opt-out even if their data is permissively licensed. This might be because the data creators did not give permission for their data to be used for training an LLM, for example.

### Governance Tool: [Am I In the Stack?](https://huggingface.co/spaces/bigcode/in-the-stack)

> This tool was developed by the [BigCode](https://www.bigcode-project.org/) team to help developers inspect The Stack dataset so they cansee whether any of their repositories have been included and might be used for training ML models. 

Reviewer: ```suggestion
This tool was developed by the [BigCode](https://www.bigcode-project.org/) team to help developers inspect The Stack dataset so they can see whether any of their repositories have been included and might be used for training ML models. 
```

If that is the case, developers can fill out a form to request to opt-out from The Stack and have their data removed for future iterations of the dataset.

> ### Governance Tool: [Have I Been Trained?](https://haveibeentrained.com/)
> This tool was developed by [Spawning](https://spawning.ai/) to help artists see if their work is a part of common datasets used to train AI art models and to opt-in or 

Reviewer: ```suggestion
### Governance Tool: [Have I Been Trained?](https://haveibeentrained.com/)

This tool was developed by [Spawning](https://spawning.ai/) to help artists see if their work is a part of common datasets used to train AI art models and to opt-in or 
```

> opt-out to future training. They maintain a "registry of non-consenting data" where individuals and companies can check to see if their data is a part of these datasets.

Reviewer: Here I'm still unclear on if the images are already under copyright (which makes it technically illegal) or if this is more a "CC license but not for ML" kind of situation.

Author: Not sure it's the right place to do a "call out" hahaha but I think it has been demonstrated that the data includes both images that are copyrighted and CC license with some use restrictions.

Not trying to delve into that issue too deeply here, but more to demonstrate that this is how orgs are thinking about how to improve transparency into what's in datasets.

Agree with you it's a big issue though.

## Data Processing

> In order to address some of the problematic aspects of the data, such as biased or harmful content or existence of personal information, some data processing should be done before the dataset is used for model training, released to a wider audience, or ideally, even before stored to a database.

Reviewer: ```suggestion
In order to address some of the problematic aspects of the data, such as biased or harmful content or existence of personal information, some data processing should be done before the dataset is used for model training, released to a wider audience, or ideally, even before stored on a database.
```

Addressing these data quality and privacy issues may be motivated by regulatory compliance like GDPR or a team’s norms and best practices.
Because this work, particularly in identifying and removing harmful content, may be contextual or reliant on the expertise of specific communities, these tasks are often performed in collaboration with community experts.

> As an example, when identifying sensitive information for removal from a dataset, a team may identify clear targets regarding an individual;s data such as usernames, passwords, and emails.

Reviewer: ```suggestion
As an example, when identifying sensitive information for removal from a dataset, a team may identify clear targets regarding an individual's data such as usernames, passwords, and emails.
```

Reviewer: Should it maybe be mentioned that many of these can be automated? i.e. validators can check for emails?

EDIT: I see this is covered in the PII section later :)

> However, a community from a specific country, or who speak a specific language, or who face a specific disability or form of discrimination may be able to identify additional criteria that should be removed to protect their community members.

Reviewer: Are there any examples of this in practice?

Author: yes i will add an example from BigScience! https://oecd.ai/en/catalogue/tools/bigscience-catalogue-of-language-data-and-resources

(pd-dg-ml)=

# Data Governance for the Machine Learning Pipeline

A Machine Learning (ML) pipeline consists of a series of activities including the collection of data, training of an ML model, and the deployment of the model into use. 
Data is integral throughout the ML process and the methods for which data is collected, annotated, processed, and shared will impact individuals and communities who may be represented in or the creators of the data, as well as data users who would like access to the data. 

> This chapter will cover examples of data governance practices for ML for different steps in the pipeline, including: 

Reviewer 2: ```suggestion
This chapter will cover examples of data governance practices for ML for different steps in the pipeline, which may include but not be exclusive to:
```
Suggested wording, feel free to change! This is just to suggest the notion that DC, DM, and DP are just part of the process, but there may be more.

- Data Collection
- Data Management
- Data Processing

(pd-dg-ml-dc)=
## Data Collection

> Many ML models are trained using datasets collected by a research team, which may be proprietary, or by using an open dataset that is available for download (sometimes with restrictions on its usage, such as only available for academic researchers).

Reviewer 3: ```suggestion
Many ML models are trained using datasets collected by a research team, which may be proprietary, or by using an open dataset that is available for download (sometimes with restrictions on its usage, such as only being available for use in academic settings).
```

Suggested change

> Many ML models are trained using datasets collected by a research team, which may be proprietary, or by using an open dataset that is available for download (sometimes with restrictions on its usage, such as only available for academic researchers).

Reviewer 2: ```suggestion
Many ML models are trained using datasets collected by a research team, which may be proprietary, or they may use an open dataset that is available for download (sometimes with restrictions on its usage, such as only available for academic researchers).
```

The deep learning (DL) family of models, in particular, rely on massive corpuses of data such as text, code, images, sound, and other media.

> The process of data collection depends on the type and volume of data required and sources for acquisition.

Reviewer 2: Perhaps a few examples might be good to cite here? I.e. a health project that requires analysing historical health records as opposed to media sentiment analysis on public opinion for elections, which would analyse social media?

> Many DL models rely on Internet data due to the sheer volume of digital content that is available on the web. 

Reviewer 2: Is there a reason why you use the term "internet data" as opposed to "data scraped from the internet"?

For example, [ImageNet datasets](https://www.image-net.org/about.php) are sourced from web images from image hosting websites like Flickr, and [LAION datasets](https://laion.ai/faq/) come from web crawling sources like Common Crawl.
These methods of data collection through web scraping have raised issues regarding data quality and bias due to the nature of using uncurated Internet data, as well as issues of data rights as web scraped data sets often include copyrighted data or data obtained without consent that may be in violation of the data license and terms of use.
To address issues related to respecting data licenses, some ML teams are choosing to only use open source or permissively licensed data.

> This can be challenging to accomplish with data that does not have explicit licenses, or for data scraped for the web which may not have the license included in the metadata.

Reviewer 3: ```suggestion
This can be challenging to accomplish with data that does not have explicit licenses, or for data scraped from the web which may not have the license included in the metadata.
```

### Governance Tool: [Datasheets for Datasets](https://www.microsoft.com/en-us/research/uploads/prod/2019/01/1803.09010.pdf)

Datasheets are resources accompanying datasets that can provide key information about the motivation to create the dataset, 
the data collection process, the dataset composition and pre-processing, and legal and ethical considerations when using the data.
Datasheets help contextualise the resource and are an opportunity for dataset creators to share any concerns, suggestions,
and other advice regarding responsible use of the dataset.
The [Datasheets for Datasets paper](https://www.microsoft.com/en-us/research/uploads/prod/2019/01/1803.09010.pdf) contains a list of example questions that data creators should ask themselves and document in their datasheet.

(pd-dg-ml-dm)=
## Data Management

Once data collection is complete, challenges emerge around data storage, access, and making changes to the dataset.
Because state-of-the-art ML datasets are very large, they tend to be stored in a central location, often a cloud provider database, and made available for download under certain circumstances.

> In addition to creating a governance structure for the opening or gating of access of the dataset to certain users, there is also the challenge of creating a governance structure to allow data creators to request to opt-out their data from the dataset, if the team is interested in providing that capability.

Reviewer 3: ```suggestion
In addition to creating a governance structure for managing access to the dataset to certain users, there is also the challenge of creating a governance structure to allow data creators to request to opt-out their data from the dataset, if the team is interested in providing that capability.
```

> Because opting out occurs after a dataset has already been created, it will not remove the data from original versions of the dataset, but can facilitate removal from future versions.

Reviewer 3: Do we mean dataset or model here?

Author: Dataset! The thinking is that once a dataset exists you can't effectively "remove" data from it as it probably already exists/has been downloaded in multiple places

Reviewer: Is a management service like [DagsHub](https://dagshub.com/) germane to this discussion?

### Governance Tool: [Have I Been Trained?](https://haveibeentrained.com/)

This tool was developed by [Spawning](https://spawning.ai/) to help artists check if their work is a part of common datasets used to train AI art models and to opt-in or 
opt-out to future training. They maintain a "registry of non-consenting data" where individuals and companies can check to see if their data is a part of these datasets.

(pd-dg-ml-dp)=
## Data Processing

In order to address some of the problematic aspects of datasets, such as biased or harmful content or existence of personal information, some data processing should be done before the dataset is used for model training, released to a wider audience, or ideally, even before stored on a database.
Addressing these data quality and privacy issues may be motivated by regulatory compliance like GDPR or a team’s norms and best practices.

> Because this work, particularly identifying and removing harmful content, may be contextual or reliant on the expertise of specific communities, these tasks are often performed in collaboration with community experts.

Reviewer 3: ```suggestion
Because this work, particularly identifying and removing harmful content, may be contextual or reliant on the expertise of specific communities, these tasks should be performed in collaboration with community experts.
```

don't want to give them impression that most companies actually do this...

The deep learning (DL) family of models, in particular, relies on massive corpuses of data such as text, code, images, sound, and other media.
The process of data collection depends on the type and volume of data required and sources for acquisition.
For example, a project that uses patient health record data versus a project that uses a dataset of millions of social media posts will require different processes for gaining access to data as well as actually collecting and storing the data.
Many DL models rely on data scraped from the internet due to the sheer volume of digital content that is available on the web. 
For example, [ImageNet datasets](https://www.image-net.org/about.php) are sourced from web images from image hosting websites like Flickr, and [LAION datasets](https://laion.ai/faq/) come from web crawling sources like Common Crawl.
These methods of data collection through web scraping have raised issues regarding data quality and bias due to the nature of using uncurated Internet data, as well as issues of data rights as web scraped data sets often include copyrighted data or data obtained without consent that may be in violation of the data license and terms of use.

> To address issues related to respecting data licenses, some ML teams are choosing to only use open source or permissively licensed data.

Reviewer: Even among open source data, there are incompatible licenses, I'm not sure how much needs to be covered here, maybe just:

```suggestion
To address issues related to respecting data licenses, ML teams should be mindful of using data only when compatible with the license, including open source or permissively licensed data.
```
